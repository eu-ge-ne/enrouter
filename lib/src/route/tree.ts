import type { Route } from "./mod.js";
//@ts-ignore
import routeTree from "virtual:enrouter";

export function getRouteTree(): Route {
  return routeTree;
}
