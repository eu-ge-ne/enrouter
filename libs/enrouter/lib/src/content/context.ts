import { createContext, useContext } from "react";

import type { Content } from "./mod.js";

const ContentContext = createContext<Content>({
  match: {
    route: {
      path: "",
      test: {
        keys: [],
        pattern: new RegExp(""),
      },
      modules: [],
      loaded: false,
      elements: {},
    },
    location: "",
    isFull: false,
    params: {},
  },
});

export const ContentProvider = ContentContext.Provider;

export function useContent(): Content {
  return useContext(ContentContext);
}
