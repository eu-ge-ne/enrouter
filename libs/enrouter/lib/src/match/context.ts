import { createContext, useContext } from "react";

import type { Match } from "./mod.js";

const MatchContext = createContext<Match>({
  route: {
    path: "",
    test: { keys: [], pattern: new RegExp("") },
    modules: [],
    loaded: false,
    elements: {},
  },

  isFull: false,
  location: "",
  params: {},
});

export const MatchProvider = MatchContext.Provider;

export function useMatch(): Match {
  return useContext(MatchContext);
}
