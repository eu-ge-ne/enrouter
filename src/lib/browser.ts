export function pushHistory(to: string) {
  window.history.pushState({}, "", to);
}

export function assignLocation(to: string) {
  window.location.assign(to);
}
