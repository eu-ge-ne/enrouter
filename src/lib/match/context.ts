import { createContext, useContext } from "react";

import type { Match } from "./match.js";

const MatchesContext = createContext<Match[]>([]);
const MatchIndexContext = createContext<number>(-1);

export const MatchesProvider = MatchesContext.Provider;
export const MatchIndexProvider = MatchIndexContext.Provider;

export interface MatchContext {
  matches: Match[];
  matchIndex: number;
  match: Match | undefined;
  firstMatch: Match | undefined;
  nextMatch: Match | undefined;
  lastMatch: Match | undefined;
  voidMatch: Match | undefined;
  isExactMatch: boolean;
}

export function useMatchContext(): MatchContext {
  const matches = useContext(MatchesContext);
  const matchIndex = useContext(MatchIndexContext);

  const match = matches[matchIndex];
  const firstMatch = matches[0];
  const nextMatch = matches[matchIndex + 1];
  const lastMatch = matches.at(-1);
  const voidMatch = matches.findLast((x) => x.route?.elements._void);

  const isExactMatch = Boolean(lastMatch?.route);

  return {
    matches,
    matchIndex,
    match,
    firstMatch,
    nextMatch,
    lastMatch,
    voidMatch,
    isExactMatch,
  };
}

export function useMatch(path: string): Match | undefined {
  return useContext(MatchesContext).find((x) => x.route?.path === path);
}
