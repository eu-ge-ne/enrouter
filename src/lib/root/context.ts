import { type ComponentType, createContext, useContext } from "react";

export interface RootParams {
  root?: ComponentType;
  fallback?: Record<string, ComponentType>;
}

const RootParamsContext = createContext<RootParams>({});

export const RootParamsProvider = RootParamsContext.Provider;

export function useRootParams(): RootParams {
  return useContext(RootParamsContext);
}
