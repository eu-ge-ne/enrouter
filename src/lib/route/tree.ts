import type { Route } from "./route.js";
//@ts-ignore
import routeTree from "virtual:enrouter";

export function getRouteTree(): Route {
  return routeTree;
}
