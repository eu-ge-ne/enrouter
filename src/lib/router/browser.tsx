import { type ReactNode, useState, useCallback, useEffect } from "react";

import type { Match } from "#lib/match/mod.js";
import { logger } from "#lib/debug.js";
import { matchLocation } from "#lib/match/location.js";
import { NavigateProvider } from "#lib/navigate/mod.js";
import { LocationProvider } from "#lib/location/mod.js";
import { MatchProvider } from "#lib/match/context.js";
import * as browser from "#lib/browser/mod.js";
import { Root } from "./root.js";

const log = logger("router/browser");

export interface BrowserRouterProps {
  match: Match | undefined;
}

export function BrowserRouter(props: BrowserRouterProps): ReactNode {
  const [location, setLocation] = useState(window.location.pathname);
  const [match, setMatch] = useState(props.match);

  const navigate = useCallback(async (to: string) => {
    let nextMatch: Match | undefined;

    try {
      nextMatch = await matchLocation(to);
    } catch (err) {
      log("matchLocation error %o", err);
    }

    setLocation(to);
    setMatch(nextMatch);

    if (nextMatch) {
      browser.pushHistory(to);
    } else {
      browser.assignLocation(to);
    }

    log("Navigated to %s", to);
  }, []);

  const handlePopState = useCallback(async (e: PopStateEvent) => {
    log("handlePopState %o", e);

    const to = window.location.pathname;
    setLocation(to);

    setMatch(await matchLocation(to));
  }, []);

  useEffect(() => {
    window.addEventListener("popstate", handlePopState);
    return () => window.removeEventListener("popstate", handlePopState);
  }, [handlePopState]);

  return (
    <NavigateProvider value={navigate}>
      <LocationProvider value={location}>
        <MatchProvider value={match}>
          <Root />
        </MatchProvider>
      </LocationProvider>
    </NavigateProvider>
  );
}
