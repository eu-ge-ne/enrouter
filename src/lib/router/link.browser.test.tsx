import type { FC, PropsWithChildren } from "react";
import { describe, test, expect, vi } from "vitest";
import { render } from "vitest-browser-react";
import { userEvent } from "@vitest/browser/context";
import * as regexparam from "regexparam";

import type { Route } from "#lib/route/mod.js";
import type { Match } from "#lib/match/mod.js";
import { getRouteTree } from "#lib/route/tree.js";
import { useLink } from "#lib/link/mod.js";
import { Outlet } from "#lib/outlet/mod.js";
import { BrowserRouter } from "./browser.js";

const wrapperId = "test-wrapper";

const wrapper: FC<PropsWithChildren> = ({ children }) => (
  <div data-testid={wrapperId}>{children}</div>
);

vi.mock(import("#lib/route/tree.js"), () => ({
  getRouteTree: vi.fn(),
}));

describe("router", () => {
  describe("BrowserRouter", () => {
    test("link", async () => {
      function TestLink() {
        const props = useLink("/abc");
        return <a {...props}>link to /abc</a>;
      }

      const route: Route = {
        path: "/",
        test: regexparam.parse("/", true),
        modules: [],
        loaded: true,
        elements: {
          _layout: {
            Root: (
              <div>
                <TestLink />
                <Outlet />
              </div>
            ),
          },
        },
        tree: [
          {
            path: "/abc",
            test: regexparam.parse("/abc", true),
            modules: [],
            loaded: true,
            elements: {
              _layout: {
                Abc: <div>abc</div>,
              },
            },
          },
        ],
      };

      vi.mocked(getRouteTree).mockReturnValueOnce(route);

      const match: Match = {
        isVoid: false,
        route,
        location: "/",
        isExact: true,
        params: {},
      };

      match.first = match;
      match.last = match;

      const screen = render(<BrowserRouter match={match} />, { wrapper });

      await userEvent.click(screen.getByRole("link"));

      await expect.element(screen.getByTestId(wrapperId)).toBeVisible();

      expect(screen.container).toMatchSnapshot();
    });
  });
});
