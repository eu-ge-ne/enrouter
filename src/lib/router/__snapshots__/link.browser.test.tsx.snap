// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`router > BrowserRouter > link 1`] = `
<div>
  <div
    data-testid="test-wrapper"
  >
    <div>
      <a
        href="/abc"
      >
        link to /abc
      </a>
      <div>
        abc
      </div>
    </div>
  </div>
</div>
`;
